src/power_geopm/nrmpower_geopm.c

```diff
@@ -0,0 +1,377 @@
+/*******************************************************************************
+ * Copyright 2021 UChicago Argonne, LLC.
+ * (c.f. AUTHORS, LICENSE)
+ *
+ * This file is part of the nrm-extra project.
+ * For more info, see https://github.com/anlsys/nrm-extra
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *******************************************************************************/
+
+/* Filename: nrmpower_papi.c
+ *
+ * Description: Implements middleware between powercap, measured via PAPI,
+ *               and the NRM downstream interface. Resources detected via hwloc.
+ */
+
+#include <limits.h>
+#include <stddef.h>
+#define _GNU_SOURCE
+#include <assert.h>
+#include <errno.h>
+#include <geopm_pio.h>
+#include <geopm_topo.h>
+#include <getopt.h>
+#include <hwloc.h>
+#include <limits.h>
+#include <math.h>
+#include <sched.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <nrm.h>
+
+#include "extra.h"
+
+static int log_level = NRM_LOG_DEBUG;
+volatile sig_atomic_t stop;
+
+static nrm_client_t *client;
+static nrm_scope_t *scope;
+static nrm_sensor_t *sensor;
+
+static char *upstream_uri = "tcp://127.0.0.1";
+static int pub_port = 2345;
+static int rpc_port = 3456;
+
+#define assert__(x) for (; !(x); assert(x))
```

Comment:
why is this needed ?

Reply:
This was a "solution" to print an explanatory message on the assertion failing, but as you stated it's better to exit cleanly on the condition not being met